### ISSUE TITLE
dump to tar doesn't support directories

### ISSUE THREAD

**Issue by pschultz** [restic/restic#2319]
Output of `restic version`
--------------------------

restic 0.9.5 compiled with go1.12.4 on linux/amd64

How did you run restic exactly?
-------------------------------

    $ restic dump latest / >/dev/null

See reproducer below for full setup.

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

filesystem

Expected behavior
-----------------

restic completes without error

Actual behavior
---------------

restic exits with error

> cannot dump file: TarHeader : archive/tar: cannot encode header: USTAR cannot encode Mode=2147484157; and PAX cannot encode Mode=2147484157; and only PAX supports PAXRecords

Steps to reproduce the behavior
-------------------------------

```
export RESTIC_REPOSITORY=repo
export RESTIC_PASSWORD=asdf
restic init
mkdir -p data/A
touch data/A/b
restic backup data
restic dump latest / >/dev/null
```

Do you have any idea what may have caused this?
-----------------------------------------------

[The `tar.Header.Mode` field supports at most 21 bits](https://golang.org/pkg/archive/tar/#Format) (the field is eight bytes long and stores seven octal digits in ASCII, followed by a null terminator, for a net storage of 21 bits).

[`NodeFromFileInfo`](https://github.com/restic/restic/blob/v0.9.5/internal/restic/node.go#L88-L93) (correctly) stores 32 bits (os.ModeDir is bit 32), but [bits 22 to 32 are not unset when initializing the tar header](https://github.com/restic/restic/blob/v0.9.5/cmd/restic/cmd_dump.go#L247). This results in a mode value that cannot be represented in the PAX format.

Do you have an idea how to solve the issue?
-------------------------------------------

Almost all of the high bits all related to inode types, so an obvious solution is to mask them out in the Mode field and set the Typeflag appropriately instead. Bit 22 is os.ModeSticky, which doesn't seem to be supported in the PAX format. [I have a PR ready for this](https://github.com/classmarkets/restic/commits/2319-dump-dir-tar).

Alternatively, drop support for extended attributes for tar archives and output the GNU format, which supports 57 bits in the mode field (more than enough for restic's purposes). This also increases the maximum filename length (refer to [the table in the tar package](https://golang.org/pkg/archive/tar/#Format) for details).

It's also reasonable to support both formats and let the user choose via a flag.

* * *

I don't see how dumping a tar archive ever worked with directories. However, I find it also unlikely that this didn't come up when testing #2124. I just want to make sure that I don't see ghosts before submitting a PR.

**Comment by Kidswiss** [comment]
Hi @pschultz 

I saw this issue only just now. But this is very interesting. I just tested your steps to reproduce and I don't have this error:
```
➜ export RESTIC_REPOSITORY=repo
export RESTIC_PASSWORD=asdf
restic init
mkdir -p data/A
touch data/A/b
restic backup data
restic dump latest / >/dev/null
created restic repository 81cae86525 at repo

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.
repository 81cae865 opened successfully, password is correct
created new cache in /Users/*snip*/Library/Caches/restic
found 69 old cache directories in /Users/*snip*/Library/Caches/restic, pass --cleanup-cache to remove them

Files:           1 new,     0 changed,     0 unmodified
Dirs:            0 new,     0 changed,     0 unmodified
Added to the repo: 381 B

processed 1 files, 0 B in 0:00
snapshot 6f41888e saved

➜ restic version
restic 0.9.5 compiled with go1.13.1 on darwin/amd64
```

The MacOs version I use is built with a newer go version, so maybe it's something they fixed in the stdlib? Having a quick glance at the tar pax header code looks like they do some formatting before writing it (but haven't had time to really dive into this).

**Comment by pschultz** [comment]
Building with 1.13.3 still produces the error for me on my workstation.

I had a colleague try it on their Mac and there was no error either. Then I tried on another Linux machine and it worked fine too. 

My best guess right now is that it's related to selinux or something else about extended attributes perhaps. I'll investigate further.

**Comment by Kidswiss** [comment]
Can you post the filesystem you're using? I only tested on EXT4 for linux on Ubuntu. Maybe other filesystems handle it differently and may need the bitmasking/disabling of ext attrs (although I'd really like to support them for every fs if possible...)

**Comment by pschultz** [comment]
I tried both on ext4 and tmpfs, so nothing exotic in and of itself. Hence my pointing fingers to selinux.

    /dev/md126p3 on /home type ext4 (rw,relatime,seclabel,stripe=16)
    tmpfs on /tmp type tmpfs (rw,nosuid,nodev,seclabel)

**Comment by Kidswiss** [comment]
Hi @pschultz 

Were you successful in reproducing this on other hosts? I have encountered a customer of ours that now faces the same issues from within OpenShift. That customer is using this feature quite a lot on various OpenShift nodes, but the issue just manifested on the newest one and I'm still not sure what exactly is responsible for this.

**Comment by pschultz** [comment]
I haven't had time to debug further, unfortunately. I'm really busy at work.

I can only tell you that classmarkets/restic@ea9424fc8dcfa00e3b47ce5902ef4a7095ecfce2 fixed it for me.

**Comment by Kidswiss** [comment]
Strange I tried the same restore on the OpenShift cluster with an older container version built with my original branch doesn't have this issue.

So I guess that the actual reason that this broke may be due to some other changes between my commit and tag 0.9.5.

**Comment by Kidswiss** [comment]
Tested it with https://github.com/classmarkets/restic/commit/ea9424fc8dcfa00e3b47ce5902ef4a7095ecfce2 and can confirm that this works perfectly! Thanks for looking into that and fixing it!

But with the changes in https://github.com/restic/restic/pull/2255 the got refactored a bit to fit better into the architecture of Restic.

Having said that I'm still not sure what exactly causes this, as I too, have more than enough going on currently to dig into this deeper. But with your changes we can ensure that the mode will always be correct for the use with PAX.

**Comment by MichaelEischer** [comment]
With newer go version, tar can also write the GNU format, which is the only one that able to encode the value of `node.Mode`.

For an explanation of what's wrong, take a look at `FileInfoHeader` at https://golang.org/src/archive/tar/common.go?s=22088:22153#L619 . Which just only uses the permission bits (0777) directly and converts the rest. Turns out the file type constants in go are different from the standard unix ones. After fixing this it's also possible to open the tar files with 7-zip on Windows.

**Comment by copart** [comment]
this issue is causing 0.10.0 to not pass tests after building on Fedora rawhide (all architectures).

Rawhide is current go v1.15.3

```
github.com/restic/restic/internal/dump
--- FAIL: TestWriteTar (0.05s)
    --- FAIL: TestWriteTar/multiple_files_and_folders_in_root (0.02s)
        testing.go:30: using low-security KDF parameters for test
        config.go:65: disabling check of the chunker polynomial
        tar_test.go:82: chdir to /tmp/restic-test-314507641
        tar_test.go:93: WriteTar() error = TarHeader : archive/tar: cannot encode header: USTAR cannot encode Mode=2147484141; and PAX cannot encode Mode=2147484141; and only PAX supports PAXRecords
        panic.go:617: chdir back to /builddir/build/BUILD/restic-0.10.0/_build/src/github.com/restic/restic/internal/dump
FAIL
exit status 1
```

**Comment by MichaelEischer** [comment]
@copart I'm pretty sure that the test should also fail with older restic versions or a older go compiler version on that machine. Juding from the error message there are some extended filesystem attributes set for the temp directory.

You could try #3039, which should fix the test failure.

**Comment by MichaelEischer** [comment]
Fixed by #3039
